<img style="float: left; margin: 30px 15px 15px 15px;" src="https://oci02.img.iteso.mx/Identidades-De-Instancia/ITESO/Logos%20ITESO/Logo-ITESO-Principal.jpg" width="400" height="600" /> 
    
    
## <font color='navy'> Correciones Examen 02
    
### <font color='navy'> Laboratorio de Aprendizaje Estadistico

    Juan Alvaro Morales Ramirez Valadez
    

<div style="background-color: #cce5ff; padding: 15px; border-radius: 8px;">
  <h3> Teoría</h3>
</div>


**1. Explica el modelo de regresión logística para clasificación. ¿Cómo se determina el umbral de decisión?**  
> La regresión logística es un modelo estadístico pensado para clasificar datos en dos clases. Lo que hace es calcular la probabilidad de que una observación pertenezca a una categoría u otra, usando la función sigmoide como base. Esa función convierte cualquier valor en una probabilidad entre 0 y 1. Luego, se usa un umbral de decisión para traducir esa probabilidad en una clase final: si la probabilidad es mayor que el umbral, se asigna a una clase; si es menor, a la otra.

---

**2. Explica la intuición de la máquina de soporte vectorial para clasificación. ¿Cómo se determina qué modelo es mejor? ¿Cuál es la mayor diferencia que tiene contra un modelo de regresión logística?**  
> Una SVM busca separar los datos en distintas clases mediante un hiperplano llamado margen que los divida de la mejor forma posible. Lo importante es que trata de maximizar la distancia entre ese hiperplano y los puntos más cercanos de cada clase, llamados vectores de soporte.
Para decidir qué modelo es mejor,se evalua el modelo con CV y con sus metricas finales.
La diferencia más clara con la regresión logística es que la SVC no se basa en probabilidades, sino en encontrar la frontera óptima entre clases, mientras que la regresión logística sí trabaja con probabilidades.

---

**3. ¿Cuáles son los componentes principales en un MLP para clasificación? Dibuja un ejemplo y señaliza.**  
> > Un MLP (Perceptrón Multicapa) se compone de varias capas de neuronas:  
>  
> - **Capa de entrada:** recibe las variables independientes o características.  
> - **Capas ocultas:** cada neurona realiza algo parecido a una regresión logística, pero con funciones de activación no lineales (como **ReLU**), lo que permite al modelo captar relaciones más complejas.  
> - **Capa de salida:** genera el resultado final, como una probabilidad o una clase.  
>  
> *(Si se dibujara, se mostraría una capa de entrada conectada a una o más capas ocultas, y luego a una capa de salida, con flechas indicando el flujo de información.)*


---

**4. ¿Cuál es el procedimiento a seguir cuando los datos no son linealmente separables en una SVC?**  
> Se tieen que utilizar un proceso llamado kernel que lo que hace es transformar el espacio o dimension de mis datos para que puedan ser separables o explicables por el modelo de clasificación

---

**5. Describe qué es un hiperparámetro. ¿Por qué es importante ajustarlos? Da dos ejemplos de hiperparámetros.**  
> Un hiperparametro es toda aquella variable que puede tener el modelo para su fincionamiento que tu puedes elegir su valor para poder buscar maximizar el rendimiento del modelo

---

**6. Dibuja un diagrama de flujo para describir el proceso de optimización Bayesiana.**  
> El proceso comienza evaluando algunos puntos aleatorios de los hiperparámetros mediante un proceso Gaussiano para aproximar la función objetivo. Luego:
>
>1.- Se estima la función objetivo.
>
>2.- Se elige si se quiere maximizar, minimizar o reducir la incertidumbre.
>
>3.- Se selecciona el siguiente punto a evaluar.
>
>4.- Se actualiza el modelo con esa nueva información.
>
>Y Este ciclo se repite hasta encontrar la combinación óptima de hiperparámetros.

---

**7. ¿Qué es la curva ROC y cómo se usa para evaluar el desempeño de un modelo?**  
> La curva ROC muestra el desempeño de un clasificador para distintos umbrales de decisión.  
> En el eje Y se coloca el TPR (recall o sensibilidad) y en el eje X el FPR (1 - especificidad).  
> Permite ver cómo cambia la sensibilidad frente a los falsos positivos.  
>  
> Se usa su integral o el área bajo la curva (AUC), que nos ayuda a ver qué tan bien distingue el modelo los casos positivos de los negativos en una escala de 0 a 1, donde 1 es el mejor resultado posible.

---

**8. Describe un espacio de Hilbert.**  
> Basicamente es un kernel. Es un espacio donde se puede extender a infinctas dimensiones para que tus datos sean explicables por el modleo de clasificacion.

---

**9. ¿Qué significa que una función de costo sea convexa? ¿Qué beneficios hay de que un modelo tenga una función de costo convexa?**  
> Cuando una función de costo es convexa, significa que tiene una única solución óptima global, sin mínimos locales que confundan al modelo. Esto facilita encontrar la mejor solución posible a un modelo..

---

**10. Piensa en los 3 modelos aprendidos en este parcial: ¿En qué situaciones usarías cada uno y por qué?**  
> **Regresión logística**: cuando los datos son linealmente separables y se busca una interpretación probabilística.
>
>**SVC**: cuando se quiere una frontera clara entre clases y se dispone de datos que podrían requerir kernels para separarse.
>
>**MLP**: cuando las relaciones entre las variables son muy complejas y no lineales, y se necesita un modelo más flexible para captarlas.


<div style="background-color: #cce5ff; padding: 15px; border-radius: 8px;">
  <h3> Práctica</h3>
</div>


# Parte 1

Utiliza optimización bayesiana para intentar encontrar el mínimo global de la siguiente función:


donde las 3 variables están acotadas en [0, 1]. Crea 5 muestras iniciales e itera 15 veces para optimizar.

In [16]:
import numpy as np
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, ConstantKernel as C
import warnings
warnings.filterwarnings("ignore")

In [18]:
# Definimos la función objetivo
def f(x, y, z):
    return ((6*x - 2)**2) * np.sin(12*x - 4) + ((6*y - 2)**2) * np.cos(12*y - 4) + ((6*z - 2)**2) * np.sin(12*z - 4)

In [20]:
#  Crear 5 muestras iniciales en el dominio [0,1]
np.random.seed(42)
X = np.random.rand(5, 3)  # 5 puntos, 3 variables
y = np.array([f(x[0], x[1], x[2]) for x in X])

In [22]:
# Definir el modelo de proceso gaussiano
kernel = C(1.0) * RBF(length_scale=1.0)
gp = GaussianProcessRegressor(kernel=kernel, n_restarts_optimizer=5, normalize_y=True)

In [24]:
# Optimización bayesiana
for i in range(15):
    # Entrenar el modelo con los datos actuales
    gp.fit(X, y)
    
    # Generar puntos candidatos aleatorios
    X_candidates = np.random.rand(1000, 3)
    
    # Predecir valor esperado y desviación estándar
    y_pred, sigma = gp.predict(X_candidates, return_std=True)
    
    # Seleccionar el punto con menor valor predicho (exploración simple)
    next_point = X_candidates[np.argmin(y_pred)]
    
    # Evaluar la función en ese punto
    next_value = f(next_point[0], next_point[1], next_point[2])
    
    # Agregar el nuevo punto
    X = np.vstack([X, next_point])
    y = np.append(y, next_value)

In [32]:
# Resultado final
best_idx = np.argmin(y)
print(f"Mejor punto encontrado: {X[best_idx]}")
print(f"Valor mínimo estimado: {y[best_idx]:.5f}")

Mejor punto encontrado: [0.75536444 0.06020446 0.14962448]
Valor mínimo estimado: -9.65871


**Explica por qué optimización bayesiana es una elección buena para este problema en lugar de GridSearch.**

>GridSearch requiere evaluar la función en toda la combinacion de coordenadas de mis variables lo cual para este caso es computacionalmente costoso y lento.

>Mientras que la Optimización bayesiana estima dónde se encuentra el mínimo probable y explora el espacio de forma inteligente, reduciendo drásticamente el número de evaluaciones necesarias.



# Parte 2

- El dataset adidas contiene información acerca de productos y la calificación que se les dio por usuarios. Utiliza regresión logística para conseguir un modelo que nos ayude a decidir si un producto será clasificado como bueno (average_rating>=4.3).

- Encuentra la significancia estadística de los factores utilizados.

- Cross-validation a utilizar: Train-test.

- Métrico a utilizar: AUC. Explica el significado el score que arroje tu modelo.

In [59]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score
import numpy as np
import pandas as pd
from scipy.stats import norm

In [61]:
df = pd.read_csv("adidas.csv")
df = df.dropna().drop_duplicates()
df.head(2)

,url,name,sku,selling_price,original_price,currency,availability,color,category,source,source_website,breadcrumbs,description,brand,images,country,language,average_rating,reviews_count,crawled_at
15,https://www.adidas.com/us/essentials-loose-log...,Essentials Loose Logo Tank Top,H07758,20,$25,USD,InStock,Purple,Clothing,adidas United States,https://www.adidas.com,Women/Clothing,What's on the agenda? Whether it's a quick lun...,adidas,"https://assets.adidas.com/images/w_600,f_auto,...",USA,en,4.8,116,2021-10-23 17:50:19.626607
16,https://www.adidas.com/us/essentials-loose-log...,Essentials Loose Logo Tank Top,H07757,20,$25,USD,InStock,Pink,Clothing,adidas United States,https://www.adidas.com,Women/Clothing,What's on the agenda? Whether it's a quick lun...,adidas,"https://assets.adidas.com/images/w_600,f_auto,...",USA,en,4.8,116,2021-10-23 17:50:19.814818


In [63]:
df.describe()

,selling_price,average_rating,reviews_count
count,829.000000,829.000000,829.000000
mean,52.911942,4.608323,433.943305
std,31.151542,0.292690,1239.690004
min,9.000000,1.000000,1.000000
25%,28.000000,4.500000,20.000000
50%,48.000000,4.700000,72.000000
75%,68.000000,4.800000,319.000000
max,240.000000,5.000000,11750.000000


In [65]:
df.columns

Index(['url', 'name', 'sku', 'selling_price', 'original_price', 'currency',
       'availability', 'color', 'category', 'source', 'source_website',
       'breadcrumbs', 'description', 'brand', 'images', 'country', 'language',
       'average_rating', 'reviews_count', 'crawled_at'],
      dtype='object')

In [67]:
# Limpiar columnas numéricas
df["selling_price"] = df["selling_price"].replace('[\$,]', '', regex=True).astype(float)
df["original_price"] = df["original_price"].replace('[\$,]', '', regex=True).astype(float)
df["reviews_count"] = df["reviews_count"].fillna(0).astype(float)

# Crear variable discount
df["discount"] = df["original_price"] - df["selling_price"]

# Variable objetivo
df["is_good"] = (df["average_rating"] >= 4.3).astype(int)

# Columnas numéricas y categóricas
num_columns = ["reviews_count", "discount"]
cat_columns = ["category", "brand", "availability"]

X = df[num_columns + cat_columns].fillna("Unknown")
y = df["is_good"]

In [69]:
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Preprocesamiento 
numerical_transformer = StandardScaler()
categorical_transformer = OneHotEncoder(handle_unknown='ignore', sparse_output=False, drop='first')

preprocessor = ColumnTransformer([
    ("num", numerical_transformer, num_columns),
    ("cat", categorical_transformer, cat_columns)
])

# Pipeline
pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", LogisticRegression(max_iter=2000))
])

In [73]:
# Entrenar modelo
pipeline.fit(X_train, y_train)

# Predecir y calcular AUC
y_pred_prob = pipeline.predict_proba(X_test)[:,1]
auc = roc_auc_score(y_test, y_pred_prob)
print(f"AUC del modelo: {auc:.3f}")

AUC del modelo: 0.716


El AUC  mide la capacidad del modelo para diferenciar entre productos buenos y no buenos. Un AUC de 0.716 significa que, si tomamos aleatoriamente un producto bueno y uno no bueno, el modelo tiene un 71.6% de probabilidad de asignar una probabilidad más alta al producto bueno que al malo.

En otras palabras, el modelo predice mejor que el azar (50%), pero todavía hay margen de mejora: cuanto más se acerque el AUC a 1, más preciso será para clasificar correctamente los productos según su rating.

In [75]:
# Significancia estadística 
# Matriz procesada
X_train_num_scaled = pipeline.named_steps["preprocessor"].named_transformers_["num"].transform(X_train[num_columns])
X_train_cat_encoded = pipeline.named_steps["preprocessor"].named_transformers_["cat"].transform(X_train[cat_columns])
X_train_processed = np.hstack([X_train_num_scaled, X_train_cat_encoded])

# Matriz de diseño con intercepto
X_design = np.column_stack([np.ones(X_train_processed.shape[0]), X_train_processed])

# Logit
modelo = pipeline.named_steps["model"]
intercepto = modelo.intercept_[0]
coeficientes = modelo.coef_[0]
logit = intercepto + np.dot(X_train_processed, coeficientes)

# Varianza, SE, z y p-valor
p = 1 / (1 + np.exp(-logit))
V = np.diagflat(p * (1 - p))
cov = np.linalg.inv(X_design.T @ V @ X_design)
se = np.sqrt(np.diag(cov))
z = np.append(intercepto, coeficientes) / se
p_values = 2 * (1 - norm.cdf(abs(z)))

# Nombres de columnas
cat_feature_names = pipeline.named_steps["preprocessor"].named_transformers_["cat"].get_feature_names_out(cat_columns)
all_feature_names = ["Intercepto"] + num_columns + list(cat_feature_names)

# DataFrame de resultados
results = pd.DataFrame({
    "Variable": all_feature_names,
    "Coeficiente": np.append(intercepto, coeficientes),
    "Error estándar": se,
    "Z": z,
    "p-valor": p_values
})

print("Significancia estadística de los factores:")
results

Significancia estadística de los factores:


,Variable,Coeficiente,Error estándar,Z,p-valor
0,Intercepto,3.193787,0.664125,4.809014,0.000002
1,reviews_count,0.774452,0.388079,1.995603,0.045977
2,discount,-0.210337,0.130385,-1.613203,0.106700
3,category_Clothing,-1.148438,0.679854,-1.689242,0.091173
4,category_Shoes,-1.332908,0.690598,-1.930077,0.053597
5,availability_OutOfStock,-0.204092,1.392532,-0.146562,0.883478


El análisis muestra que la variable más relevante es reviews_count: a mayor cantidad de reseñas, mayor la probabilidad de que un producto sea considerado bueno, y esto es estadísticamente significativo. Las demás variables como discount, las categorías (Clothing, Shoes) y availability_OutOfStock— no resultan significativas, aunque las categorías muestran una ligera tendencia negativa frente a la categoría base. En resumen, el modelo indica que la cantidad de reseñas es el factor que realmente influye en la clasificación de un producto como bueno, mientras que los descuentos, la categoría y la disponibilidad no aportan evidencia clara de efecto.

# Parte 3

- Utiliza el modelo de tu elección, con los hiperparámetros y factores de tu elección para obtener el mejor predictor que puedas conseguir para decidir si un paciente del dataset padece diabetes.

- Cross-validation a utilizar: K-folds.

- Métrico a utilizar: Precisión.

In [36]:
import pandas as pd
df = pd.read_csv("diabetes.csv")

In [38]:
df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [40]:
df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


In [42]:
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer 
from sklearn.pipeline import Pipeline # flujo trabajo
from sklearn.neural_network import MLPClassifier # Modelo red neuronal
from sklearn.model_selection import StratifiedKFold, cross_val_score # k-folds cross validation
import numpy as np 
from sklearn.gaussian_process import GaussianProcessRegressor # Gaussiano
from sklearn.gaussian_process.kernels import RBF # RBF

In [44]:
# Elegir columnas
num_columns = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age',]

cat_columns = []

target = ["Outcome"]

# Separar en x y
x = df[num_columns + cat_columns]

y = df[target].values.ravel()

# CV
kfolds = 10

In [46]:
numerical_transformer = StandardScaler() # Estandariza los datos

preprocessor = ColumnTransformer(
    transformers = [
        ("num", numerical_transformer, num_columns)
    ]
)

In [48]:
# Funcion optimiacion objetivo
def foo(hidden_layers):
    model = MLPClassifier(hidden_layer_sizes=hidden_layers, random_state=10, max_iter=10_000)
    
    pipeline = Pipeline([
        ("scaler", preprocessor),
        ("model", model)
    ])

    scores = cross_val_score(pipeline, x, y, scoring="precision", cv=kfolds)
    return np.mean(scores)

In [50]:
# Valores iniciales
x1 = [10,5]
x2 = [15,10]
x3 = [25,15]

X = np.vstack([x1,x2,x3])

# Regresion Gaussiana Inicial
y_gp = []
for i in X:
    score = foo(i)
    y_gp.append(score)

kernel = RBF(length_scale=5)
gp = GaussianProcessRegressor(kernel = kernel, random_state=10)
gp.fit(X,y_gp)

GaussianProcessRegressor(kernel=RBF(length_scale=5), random_state=10)

In [51]:
# Valores aleatorios a probar 
x_capa1 = np.random.uniform(10,30, 1000).reshape(-1, 1)
x_capa2 = np.random.uniform(5, 25, 1000).reshape(-1, 1)
x_gp = np.hstack([x_capa1, x_capa2]).round()

# Optimizacion Bayesiana con 15 iteraciones
for i in range(15):
    gp.fit(X,y_gp)

    y_pred, sigma = gp.predict(x_gp, return_std=True)
    y_upper = y_pred+ 1.96 * sigma
    
    # Elegir el punto score maximo AUC
    max_pred_index = np.argmax(y_upper) 
    max_pred_x_value = x_gp[max_pred_index, :]

    # Evaluar f(x) en el nuevo punto
    y_new = foo(max_pred_x_value.astype(int))

    X = np.vstack([X, max_pred_x_value])
    print(max_pred_x_value)
    y_gp = np.append(y_gp, y_new)

[29.  5.]
[12. 25.]
[29. 25.]
[10. 16.]
[10. 23.]
[21. 25.]
[30. 12.]
[10.  8.]
[16. 20.]
[21.  5.]
[25.  9.]
[13.  5.]
[30. 19.]
[13.  5.]
[13.  5.]


In [52]:
# Resultados
best_idx = np.argmax(y_gp)
best_params = X[best_idx]
best_score = y_gp[best_idx]

print(f"Mejor configuración: {best_params}")
print(f"Mejor Precision: {best_score:.4f}")
print(f"Precision promedio: {np.mean(y_gp):.4f}")
print(f"Desviación estándar: {np.std(y_gp):.4f}")

Mejor configuración: [13.  5.]
Mejor Precision: 0.6689
Precision promedio: 0.6233
Desviación estándar: 0.0294
